From 5a87bc26b2ad2e8635d4372527188e8a0e0fc269 Mon Sep 17 00:00:00 2001
From: Rob Armstrong JR <ra@xilinx.com>
Date: Fri, 24 May 2013 02:15:52 +0200
Subject: [PATCH 023/793] char: xdevcfg: Ensure PCAP loopback is disabled

Some versions of the Zynq first stage boot loader enable PCAP loopback
during boot regardless of whether or not the boot image includes PL
configuration. This behavior only appears in certain boot modes (notably
QSPI boot). Attempting to configure the PL with the loopback bit set
will result in timeouts and will prevent successful configuration.

In order to avoid this problem, and to avoid dependency on the version
of the FSBL used to boot the system, ensure that the loopback enable bit
is cleared when loading the driver.

Signed-off-by: Rob Armstrong <ra@xilinx.com>
Signed-off-by: Michal Simek <michal.simek@xilinx.com>
---
 drivers/char/xilinx_devcfg.c |    9 +++++++++
 1 files changed, 9 insertions(+), 0 deletions(-)

diff --git a/drivers/char/xilinx_devcfg.c b/drivers/char/xilinx_devcfg.c
index e7b1de4..1ea33ee 100644
--- a/drivers/char/xilinx_devcfg.c
+++ b/drivers/char/xilinx_devcfg.c
@@ -52,6 +52,7 @@ static DEFINE_MUTEX(xdevcfg_mutex);
 #define XDCFG_DMA_SRC_LEN_OFFSET	0x20 /* DMA Source Transfer Length */
 #define XDCFG_DMA_DEST_LEN_OFFSET	0x24 /* DMA Destination Transfer */
 #define XDCFG_UNLOCK_OFFSET		0x34 /* Unlock Register */
+#define XDCFG_MCTRL_OFFSET		0x80 /* Misc. Control Register */
 
 /* Control Register Bit definitions */
 #define XDCFG_CTRL_PCFG_PROG_B_MASK	0x40000000 /* Program signal to
@@ -80,6 +81,9 @@ static DEFINE_MUTEX(xdevcfg_mutex);
 						    *  including: DAP_En,
 						    *  DBGEN,NIDEN, SPNIEN */
 
+/* Miscellaneous Control Register bit definitions */
+#define XDCFG_MCTRL_PCAP_LPBK_MASK	0x00000010 /* Internal PCAP loopback */
+
 /* Status register bit definitions */
 #define XDCFG_STATUS_PCFG_INIT_MASK	0x00000010 /* FPGA init status */
 
@@ -1949,6 +1953,11 @@ static int __devinit xdevcfg_drv_probe(struct platform_device *pdev)
 				XDCFG_CTRL_USER_MODE_MASK |
 				ctrlreg));
 
+	/* Ensure internal PCAP loopback is disabled */
+	ctrlreg = xdevcfg_readreg(drvdata->base_address + XDCFG_MCTRL_OFFSET);
+	xdevcfg_writereg(drvdata->base_address + XDCFG_MCTRL_OFFSET,
+				(~XDCFG_MCTRL_PCAP_LPBK_MASK &
+				ctrlreg));
 
 	cdev_init(&drvdata->cdev, &xdevcfg_fops);
 	drvdata->cdev.owner = THIS_MODULE;
-- 
1.7.5.4

